File: leetcode/3068-find-the-maximum-sum-of-node-values.js
// https://leetcode.com/problems/find-the-maximum-sum-of-node-values/

/**
 * @param {number[]} nums
 * @param {number} k
 * @param {number[][]} edges
 * @return {number}
 */
var maximumValueSum = function(nums, k, edges) {
  const diff = [];
  for (const num of nums) diff.push((num ^ k) - num);
  diff.sort((a, b) => (b - a));

  let maximum = nums.reduce((curr, acc) => (curr + acc), 0);
  for (let i = 0; i < diff.length - 1; i += 2) {
    if (diff[i] + diff[i+1] <= 0) break;
    maximum += (diff[i] + diff[i+1]);
  }

  return maximum;
};

File: leetcode/3068-find-the-maximum-sum-of-node-values.java
// https://leetcode.com/problems/find-the-maximum-sum-of-node-values/

import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

class Solution {
	public long maximumValueSum(int[] nums, int k, int[][] edges) {
		List<Integer> diff = new ArrayList<>();
		for (int i = 0; i < nums.length; i += 1) diff.add((nums[i] ^ k) - nums[i]);
		Collections.sort(diff);
		int l = 0;
		int r = diff.size() - 1;
		while (l < r) {
			int tmp = diff.get(l);
			diff.set(l, diff.get(r));
			diff.set(r, tmp);
			l += 1;
			r -= 1;
		}

		long maximum = 0;
		for (int i = 0; i < nums.length; i += 1) maximum += nums[i];
		for (int i = 0; i < diff.size() - 1; i += 2) {
			if (diff.get(i) + diff.get(i+1) <= 0) break;
			maximum += diff.get(i) + diff.get(i+1);
		}

		return maximum;
	}
}
